modules: lvgl: fix multiple LVGL input nodes

gschwaer opened the pull request:
The input callback does not use an instance specific name, so we can only every have one instance of this type. This issue was evident by a build error, e.g., when using "zephyr,lvgl-keypad-input" nodes twice:

```
  /path/to/include/zephyr/input/input.h:159:37: error: redefinition of '_input_callback__lvgl_keypad_process_event'
    159 |                                     _input_callback__##_callback)
        |                                     ^~~~~~~~~~~~~~~~~
```

Called from

```C
#define LVGL_INPUT_DEFINE(inst, type, msgq_size, process_evt_cb)                                   \
        INPUT_CALLBACK_DEFINE(LVGL_INPUT_DEVICE(inst), process_evt_cb,                             \
                              (void *)DEVICE_DT_INST_GET(inst));                                   \
        K_MSGQ_DEFINE(lvgl_input_msgq_##type##_##inst, sizeof(lv_indev_data_t), msgq_size, 4)
```

gschwaer committed:
modules: lvgl: fix multiple LVGL inputs

The LVGL input callback does not use an instance specific name, so we
can only ever have one instance of this type. This issue was evident by
a build error, e.g., when using "zephyr,lvgl-keypad-input" nodes twice:

  error: redefinition of '_input_callback__lvgl_keypad_process_event'

Signed-off-by: Gero Schwäricke <gero.schwaericke@posteo.de>

gschwaer commented:
CC @fabiobaltieri

gschwaer commented:
This may also be broken for `INPUT_KEYMAP_DEFINE`, `INPUT_LONGPRESS_DEFINE`, `KSCAN_INPUT_INIT`. I didn't test that though.

faxe1008 commented:
@gschwaer Thank you very much for finding this issue and providing a fix 👍

fabiobaltieri commented on modules/lvgl/include/lvgl_common_input.h:
```diff
@@ -36,8 +36,9 @@ int lvgl_init_input_devices(void);
-			      (void *)DEVICE_DT_INST_GET(inst));                                   \
+	INPUT_CALLBACK_DEFINE_NAMED(LVGL_INPUT_DEVICE(inst), process_evt_cb,                       \
+				    (void *)DEVICE_DT_INST_GET(inst),                              \
+				    _input_callback__process_evt_cb__##inst);                      \
```
```suggestion
				    process_evt_cb__##inst);                      \
```

the macro already adds a `_input_callback__`, no need to double up on it

gschwaer replied:
Does it? I thought only `INPUT_CALLBACK_DEFINE()` does that. 🤔 

Does `INPUT_CALLBACK_DEFINE()` then duplicate the `_input_callback` prefix?

fabiobaltieri replied:
oh, hey you are right it doesn't! it should though, would you mind adding a commit to fix that while you are at this? remove the prefix from https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/input/input.h#L159 and move it up to https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/input/input.h#L140 so that the APIs are coherent

gschwaer replied:
Done. I added the change to this PR, if I should instead open a new PR, let me know.

fabiobaltieri replied:
No that's great, thanks a lot!

gschwaer committed:
input: fix missing namespace prefix for named input callbacks

Using INPUT_CALLBACK_DEFINE_NAMED would not add the _input_callback__
prefix to the given callback name. This can result in namespace clashes.

Signed-off-by: Gero Schwäricke <gero.schwaericke@posteo.de>

fabiobaltieri reviewed (dismissed):
Good stuff thank you!

gschwaer committed:
kscan: input: fix multiple kscan inputs

The kscan input callback does not use an instance specific name, so we
can only ever have one instance of this type.

Signed-off-by: Gero Schwäricke <gero.schwaericke@posteo.de>

gschwaer commented:
I realized the same could happen for the kscan input. So I fixed that as well. Sorry for the noise.

gschwaer commented on drivers/kscan/kscan_input.c:
```diff
@@ -100,18 +100,15 @@ static const struct kscan_driver_api kscan_input_driver_api = {
-			      &kscan_input_config_##index,                     \
-			      POST_KERNEL, CONFIG_KSCAN_INIT_PRIORITY,         \
-			      &kscan_input_driver_api);
+#define KSCAN_INPUT_INIT(index)                                                                    \
+	INPUT_CALLBACK_DEFINE_NAMED(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb,          \
+				    (void *)DEVICE_DT_INST_GET(index), kscan_input_cb##index);     \
+	static const struct kscan_input_config kscan_input_config_##index = {                      \
+		.input_dev = DEVICE_DT_GET(DT_INST_PARENT(index)),                                 \
+	};                                                                                         \
+	static struct kscan_input_data kscan_input_data_##index;                                   \
+	DEVICE_DT_INST_DEFINE(index, kscan_input_init, NULL, &kscan_input_data_##index,            \
+			      &kscan_input_config_##index, POST_KERNEL,                            \
+			      CONFIG_KSCAN_INIT_PRIORITY, &kscan_input_driver_api);
```
Unfortunately clang-format destroyed the diff. My only change was
```diff
- INPUT_CALLBACK_DEFINE(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb, (void *)DEVICE_DT_INST_GET(index));
+ INPUT_CALLBACK_DEFINE_NAMED(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb, (void *)DEVICE_DT_INST_GET(index), kscan_input_cb##index);
```

gschwaer commented on drivers/kscan/kscan_input.c:
```diff
@@ -100,18 +100,15 @@ static const struct kscan_driver_api kscan_input_driver_api = {
-			      &kscan_input_config_##index,                     \
-			      POST_KERNEL, CONFIG_KSCAN_INIT_PRIORITY,         \
-			      &kscan_input_driver_api);
+#define KSCAN_INPUT_INIT(index)                                                                    \
+	INPUT_CALLBACK_DEFINE_NAMED(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb,          \
+				    (void *)DEVICE_DT_INST_GET(index), kscan_input_cb_##index);    \
+	static const struct kscan_input_config kscan_input_config_##index = {                      \
+		.input_dev = DEVICE_DT_GET(DT_INST_PARENT(index)),                                 \
+	};                                                                                         \
+	static struct kscan_input_data kscan_input_data_##index;                                   \
+	DEVICE_DT_INST_DEFINE(index, kscan_input_init, NULL, &kscan_input_data_##index,            \
+			      &kscan_input_config_##index, POST_KERNEL,                            \
+			      CONFIG_KSCAN_INIT_PRIORITY, &kscan_input_driver_api);
```
Unfortunately clang-format destroyed the diff. My only change was
```diff
- INPUT_CALLBACK_DEFINE(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb, (void *)DEVICE_DT_INST_GET(index));
+ INPUT_CALLBACK_DEFINE_NAMED(DEVICE_DT_GET(DT_INST_PARENT(index)), kscan_input_cb, (void *)DEVICE_DT_INST_GET(index), kscan_input_cb_##index);
```

fabiobaltieri replied:
yeah it makes a bit of a mess, actually would it bother you too much to change this and the previous patch to minimize the diff and not use clang-format? just intend them manually and keep them on 80 columns

not a clang-format fun myself

fabiobaltieri commented on include/zephyr/input/input.h:
```diff
@@ -136,11 +136,11 @@ struct input_callback {
-		.dev = _dev,                                                   \
-		.callback = _callback,                                         \
-		.user_data = _user_data,                                       \
+#define INPUT_CALLBACK_DEFINE_NAMED(_dev, _callback, _user_data, name)                             \
+	static const STRUCT_SECTION_ITERABLE(input_callback, _input_callback__##name) = {          \
+		.dev = _dev,                                                                       \
+		.callback = _callback,                                                             \
+		.user_data = _user_data,                                                           \
```
yeah so if you could manually format and only change the lines that are needed it'd be great, less diffs to mess with, this should be

```c
#define INPUT_CALLBACK_DEFINE_NAMED(_dev, _callback, _user_data, name)         \
        static const STRUCT_SECTION_ITERABLE(input_callback,                   \
                                             _input_callback__##name) = {      \
                .dev = _dev,                                                   \
                .callback = _callback,                                         \
                .user_data = _user_data,                                       \
        }
```

fabiobaltieri reviewed (approved):
thanks!!
